### PR TITLE
Unravel cyclical dependency between OutputLibrary and TypeFactory

### PR DESCRIPTION
**Pull request by nisha-bhatia**
Fixes: https://github.com/microsoft/typespec/issues/3360

### CODE REVIEW

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/InputType.cs [hunk 1]
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+using System.Collections.Generic;
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.Input/src/InputTypes/InputType.cs:4]: nit: remove

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs [hunk 2]
```diff
@@ -31,12 +31,12 @@ internal sealed class MrwSerializationTypeProvider : TypeProvider
...
         private readonly CSharpType _iPersistableModelTInterface;
         private readonly CSharpType? _iPersistableModelObjectInterface;
-        private ModelProvider _model;
+        private TypeProvider _model;
         private InputModelType _inputModel;
         private readonly FieldProvider? _rawDataField;
         private readonly bool _isStruct;
 
-        public MrwSerializationTypeProvider(ModelProvider model, InputModelType inputModel)
+        public MrwSerializationTypeProvider(TypeProvider model, InputModelType inputModel)
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs:39]: Shouldn't this one still be ModelProvider? Especially since the second parameter is InputModelType.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/ScmTypeFactory.cs [hunk 3]
```diff
@@ -3,9 +3,11 @@
 
 using System;
 using System.Collections.Generic;
+using System.Data.SqlTypes;
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/ScmTypeFactory.cs:6]: nit: remove

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/ScmTypeFactory.cs [hunk 4]
```diff
@@ -3,9 +3,11 @@
...
 using System.IO;
 using System.Linq;
 using System.Net;
+using System.Reflection;
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/ScmTypeFactory.cs:10]: nit: remove

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs [hunk 5]
```diff
@@ -10,58 +10,60 @@ namespace Microsoft.Generator.CSharp
...
+        public IReadOnlyList<TypeProvider> Enums => _enums ??= BuildEnums();
+        public IReadOnlyList<TypeProvider> Models => _models ??= BuildModels();
         public IReadOnlyList<ClientProvider> Clients => _clients ??= BuildClients();
 
-        public IDictionary<InputEnumType, EnumProvider> EnumMappings { get; }
-        public IDictionary<InputModelType, ModelProvider> ModelMappings { get; }
+        private IDictionary<InputEnumType, TypeProvider> EnumMappings { get; }
+        private IDictionary<InputModelType, TypeProvider> ModelMappings { get; }
+
+        private readonly Lazy<(TypeProvider[] Enums, TypeProvider[] Models)> _allModels;
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs:59]: I think this code was already like this, but why do we use `Lazy` here as opposed to just calling InitializeAllModels from the ctor? Won't we always need to get initialize the models when generating a library?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs [hunk 6]
```diff
@@ -10,58 +10,60 @@ namespace Microsoft.Generator.CSharp
...
 
-        public IReadOnlyList<EnumProvider> Enums => _enums ??= BuildEnums();
-        public IReadOnlyList<ModelProvider> Models => _models ??= BuildModels();
+        public IReadOnlyList<TypeProvider> Enums => _enums ??= BuildEnums();
+        public IReadOnlyList<TypeProvider> Models => _models ??= BuildModels();
         public IReadOnlyList<ClientProvider> Clients => _clients ??= BuildClients();
 
-        public IDictionary<InputEnumType, EnumProvider> EnumMappings { get; }
-        public IDictionary<InputModelType, ModelProvider> ModelMappings { get; }
+        private IDictionary<InputEnumType, TypeProvider> EnumMappings { get; }
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs:56]: nit: can make these readonly fields now that they are private.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CSharpGen.cs [hunk 7]
```diff
@@ -33,7 +34,7 @@ public async Task ExecuteAsync()
             Directory.CreateDirectory(Path.Combine(generatedSourceOutputPath, "Models"));
             List<Task> generateFilesTasks = new();
 
-            foreach (var model in output.Models)
+            foreach (TypeProvider model in output.Models) // now is of type TypeProvider and not ModelProvider so this doesn't compile
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CSharpGen.cs:37]: Is this comment accurate?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/CSharpType.cs [hunk 8]
```diff
@@ -198,7 +198,7 @@ internal CSharpType(TypeProvider implementation, IReadOnlyList<CSharpType>? argu
         public bool IsCollection => _isCollection ??= TypeIsCollection();
         public Type FrameworkType => _type ?? throw new InvalidOperationException("Not a framework type");
         public object Literal => _literal ?? throw new InvalidOperationException("Not a literal type");
-        internal TypeProvider Implementation => _implementation ?? throw new InvalidOperationException($"Not implemented type: '{Namespace}.{Name}'");
+        public TypeProvider Implementation => _implementation ?? throw new InvalidOperationException($"Not implemented type: '{Namespace}.{Name}'");
```
Comment (JoshLove-msft) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/CSharpType.cs:201]: Is this public for overriders of OutputLibrary?

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs (hunk 2 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs:39]: I don't think we want to force customers to inherit from our ModelProvider if they want to completely write a new type of model so TypeProvider allows them to do this.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/ClientModelPlugin.cs [hunk 9]
```diff
@@ -36,15 +36,17 @@ public class ClientModelPlugin : CodeModelPlugin
...
         {
-            if (inputModel.Usage.HasFlag(InputModelTypeUsage.Json))
+            switch (inputType)
             {
-                // Add MRW serialization type provider
-                return [new MrwSerializationTypeProvider(provider, inputModel)];
+                case InputModelType inputModel:
+                    return [new MrwSerializationTypeProvider(provider, inputModel)];
+                case InputEnumType inputEnum:
+                    return [EnumProvider.Create(inputEnum)];
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/ClientModelPlugin.cs:46]: I don't think this should return an EnumProvider it should be returning the serialization that an enum needs.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs [hunk 10]
```diff
@@ -10,58 +10,60 @@ namespace Microsoft.Generator.CSharp
...
-        private IReadOnlyList<ModelProvider>? _models;
+        private IReadOnlyList<TypeProvider>? _enums;
+        private IReadOnlyList<TypeProvider>? _models;
         private IReadOnlyList<ClientProvider>? _clients;
 
         public OutputLibrary()
         {
-            EnumMappings = new Dictionary<InputEnumType, EnumProvider>();
-            ModelMappings = new Dictionary<InputModelType, ModelProvider>();
+            EnumMappings = new Dictionary<InputEnumType, TypeProvider>();
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs:19]: Since this is only for caching to not create the same thing twice the cache should live with the class that does the creation.

Otherwise every consumer of TypeFactory would have to implement their own caching.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CSharpGen.cs (hunk 7 above)
Comment (nisha-bhatia) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CSharpGen.cs:37]: no - it was from a previous iteration. I will remove it

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/CSharpType.cs (hunk 8 above)
Comment (nisha-bhatia) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/CSharpType.cs:201]: Yes, I think so

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs (hunk 2 above)
Comment (ArcturusZhang) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs:39]: does this really matter because this class is internal, no one could override this.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CodeModelPlugin.cs [hunk 11]
```diff
@@ -47,6 +47,22 @@ public CodeModelPlugin(GeneratorContext context)
...
+                    if (!inputEnumType.IsExtensible)
+                    {
+                        return new List<TypeProvider> { new FixedEnumSerializationProvider(EnumProvider.Create(inputEnumType)) };
+                    }
+                }
+            }
+            catch (InvalidCastException)
+            {
+            }
+            return Array.Empty<TypeProvider>();
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CodeModelPlugin.cs:65]: what is the purpose of this method?
If I quote here: https://github.com/microsoft/typespec/pull/3621/files#r1647820847, if they want to have a completely different implementation of enum (fixed enum), they should start over from `TypeProvider`, or at least from `EnumProvider`, and last time when I am implemnetation enums, there is a statement that the conversion from the enum to string and vice versa is not considered as "serialization" because this is not something that usually could be changed for example when the library supports or do not support MRW.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/FixedEnumProvider.cs [hunk 12]
```diff
@@ -67,10 +69,11 @@ public override ValueExpression ToSerial(ValueExpression enumExpression)
...
 
             // otherwise we call the corresponding extension method to convert the value
-            return new InvokeStaticMethodExpression(Serialization?.Type, $"ToSerial{ValueType.Name}", [enumExpression], CallAsExtension: true);
+            CSharpType? serializationType = SerializationProviders.FirstOrDefault()?.Type;
+            return new InvokeStaticMethodExpression(serializationType, $"ToSerial{ValueType.Name}", [enumExpression], CallAsExtension: true);
         }
 
         public override ValueExpression ToEnum(ValueExpression valueExpression)
-            => new InvokeStaticMethodExpression(Serialization?.Type, $"To{Type.Name}", [valueExpression], CallAsExtension: true);
+            => new InvokeStaticMethodExpression(SerializationProviders.FirstOrDefault()?.Type, $"To{Type.Name}", [valueExpression], CallAsExtension: true);
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/FixedEnumProvider.cs:77]: what if they override that method in TypeFactory and returns nothing?
what if they override the enum implementation so that the enum no longer carries a `To{Type.Name}` method?
I do not think the changes in this type makes much sense.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/TypeProvider.cs [hunk 13]
```diff
@@ -111,6 +114,8 @@ private TypeSignatureModifiers GetDeclarationModifiersInternal()
 
         protected virtual TypeProvider[] BuildNestedTypes() => Array.Empty<TypeProvider>();
 
+        protected virtual TypeProvider[] BuildSerializationProviders() => Array.Empty<TypeProvider>();
```
Comment (ArcturusZhang) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/TypeProvider.cs:117]: I think not every type we want to generate could have a concept of "serialization". So why do we move it here?

**Review by ArcturusZhang — COMMENTED** [review]
maybe this PR unraveled the cyclic dependency between output library and type factory by changing the implementation of `CreateCSharpType` method in TypeFactory for enums and models, but removing the cache of models and enums by default I think is not acceptable.
With this change, every time the same enum/model type shows up there should would be a type provider created for it.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CodeModelPlugin.cs (hunk 11 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CodeModelPlugin.cs:65]: This method allows someone to override what serialization type providers get used for their types both enum and model.

For model we default to no serialization, and MGC.ClientModel will override to add in MrwSerialization for models and leave enums as default.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CodeModelPlugin.cs [hunk 14]
```diff
@@ -47,6 +47,22 @@ public CodeModelPlugin(GeneratorContext context)
         /// </summary>
         /// <param name="provider">The model type provider.</param>
         /// <param name="inputModel">The input model.</param>
-        public abstract IReadOnlyList<TypeProvider> GetSerializationTypeProviders(ModelProvider provider, InputModelType inputModel);
+        public virtual IReadOnlyList<TypeProvider> GetSerializationTypeProviders(TypeProvider provider, InputType inputModel)
+        {
+            try
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/CodeModelPlugin.cs:52]: remove try catch

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs [hunk 15]
```diff
@@ -10,58 +10,50 @@ namespace Microsoft.Generator.CSharp
...
 
-        private readonly Lazy<(EnumProvider[] Enums, ModelProvider[] Models)> _allModels;
-
-        private (EnumProvider[] Enums, ModelProvider[] Models) InitializeAllModels()
+        private (TypeProvider[] Enums, TypeProvider[] Models) InitializeAllModels()
         {
             var input = CodeModelPlugin.Instance.InputLibrary.InputNamespace;
 
-            var enums = new EnumProvider[input.Enums.Count];
+            var enums = new TypeProvider[input.Enums.Count];
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs:24]: Lets just move the enum portion to BuildEnums and the model portion to BuildModels and drop the tuple.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/FixedEnumProvider.cs (hunk 12 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/FixedEnumProvider.cs:77]: Lets open an issue to remove ToEnum and ToSerial from the base EnumProvider class since these are serialization methods and should not be required by the EnumProvider they should be add ons by the SerializationProvider.  This will ensure the name lives with the SerializationProvider and this conflict can't arise.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs (hunk 2 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp.ClientModel/src/Providers/MrwSerializationTypeProvider.cs:39]: The calling signature needs to change to TypeProvider so this also needs to change.  The fact that nothing has broken is proof that ModelProvider is not special here and MrwSerializationProvider can work with any TypeProvider.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/TypeProvider.cs (hunk 13 above)
Comment (m-nash) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/TypeProvider.cs:117]: Every type can have a serialization which is why we moved this to the base class

**Comment by m-nash** [comment]
> maybe this PR unraveled the cyclic dependency between output library and type factory by changing the implementation of `CreateCSharpType` method in TypeFactory for enums and models, but removing the cache of models and enums by default I think is not acceptable. With this change, every time the same enum/model type shows up there should would be a type provider created for it.

The issue with caching is if someone overrides CreateType they can completely bypass this caching.  BuildModels in OutputLibrary is designed to run exactly one time, and within there it will iterate over the models in InputLibrary exactly one time therefore I don't think we will run into any issues.  If someone overrides down to this layer they will need to ensure the same or build their own cache.

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/FixedEnumProvider.cs (hunk 12 above)
Comment (nisha-bhatia) [reply packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/FixedEnumProvider.cs:77]: Created https://github.com/microsoft/typespec/issues/3651

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs [hunk 16]
```diff
@@ -11,51 +11,36 @@ namespace Microsoft.Generator.CSharp
...
-        public OutputLibrary()
-        {
-            _allModels = new(InitializeAllModels);
-        }
+        public IReadOnlyList<TypeProvider> Enums => BuildEnums();
+        public IReadOnlyList<TypeProvider> Models => BuildModels();
+        public IReadOnlyList<ClientProvider> Clients => _clients ??= BuildClients();
 
-        private (TypeProvider[] Enums, TypeProvider[] Models) InitializeAllModels()
+        public virtual TypeProvider[] BuildEnums()
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs:20]: these `Build*` should be protected

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs [hunk 17]
```diff
@@ -10,81 +9,59 @@ namespace Microsoft.Generator.CSharp
...
-            ModelMappings = new Dictionary<InputModelType, ModelProvider>();
-
-            _allModels = new(InitializeAllModels);
-        }
-
-        private readonly Lazy<(EnumProvider[] Enums, ModelProvider[] Models)> _allModels;
+        public IReadOnlyList<TypeProvider> Enums => _enums ??= BuildEnums();
+        public IReadOnlyList<TypeProvider> Models => _models ??= BuildModels();
+        public IReadOnlyList<ClientProvider> Clients => _clients ??= BuildClients();
+        public virtual IReadOnlyList<TypeProvider> Types => _types ??= BuildTypes();
```
Comment (m-nash) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/OutputTypes/OutputLibrary.cs:22]: the public property should not be virtual only the build methods

**Review by m-nash — APPROVED** [review]

File: packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelProvider.cs [hunk 18]
```diff
@@ -48,9 +43,16 @@ public ModelProvider(InputModelType inputModel)
                 _declarationModifiers |= TypeSignatureModifiers.Abstract;
             }
 
-            SerializationProviders = CodeModelPlugin.Instance.GetSerializationTypeProviders(this, _inputModel);
+            _isStruct = false; // this is only a temporary placeholder because we do not support to generate structs yet.
```
Comment (ShivangiReja) [inline packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Providers/ModelProvider.cs:46]: Could you please clarify why this change was made? It should be:
> _isStruct = inputModel.ModelAsStruct;
